### PR TITLE
MULE-8688 Requestor should not use selector threads only for handling…

### PR DESCRIPTION
**Pull request by dfeist**
… responses by default.

### CODE REVIEW

File: core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+public abstract class AbstractThreadingProfileProcessingStrategy implements ProcessingStrategy
+{
+
+    protected Integer maxThreads;
+    protected Integer minThreads;
+    protected Integer maxBufferSize;
+    protected Long threadTTL;
+    protected Long threadWaitTimeout;
+    protected Integer poolExhaustedAction;
+    protected ProcessingStrategy synchronousProcessingStrategy = new SynchronousProcessingStrategy();
```
Comment (andresgregoire) [inline core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java]: Is this variable needed?

File: core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+
+    protected ThreadingProfile createThreadingProfile(MuleContext muleContext)
```
Comment (andresgregoire) [inline core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java]: Fix code style in this method

File: modules/http/src/main/java/com/ning/http/client/providers/grizzly/HttpTransactionContext.java [hunk 3]
```diff
@@ -151,7 +151,7 @@ Connection getConnection() {
         return connection;
     }
 
-    AsyncHandler getAsyncHandler() {
+    public AsyncHandler getAsyncHandler() {
```
Comment (andresgregoire) [inline modules/http/src/main/java/com/ning/http/client/providers/grizzly/HttpTransactionContext.java]: Fix code style

File: modules/http/src/main/java/org/mule/module/http/internal/request/DefaultHttpRequester.java [hunk 4]
```diff
@@ -536,4 +540,8 @@ public void setMuleContext(MuleContext muleContext)
         this.muleContext = muleContext;
     }
 
+    public void setFlowConstruct(FlowConstruct flowConstruct)
```
Comment (andresgregoire) [inline modules/http/src/main/java/org/mule/module/http/internal/request/DefaultHttpRequester.java]: Add Override annotation

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ */
+public class FlowWorkManagerIOStrategy extends AbstractIOStrategy
+{
+
+    private static final FlowWorkManagerIOStrategy INSTANCE = new FlowWorkManagerIOStrategy();
+
+    private final static EnumSet<IOEvent> WORKER_THREAD_EVENT_SET =
+            EnumSet.of(IOEvent.READ, IOEvent.CLOSED);
+
+    private static final Logger logger = Grizzly.logger(FlowWorkManagerIOStrategy.class);
```
Comment (andresgregoire) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: Use slf4j for logging

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import org.glassfish.grizzly.Connection;
+import org.glassfish.grizzly.Grizzly;
+import org.glassfish.grizzly.IOEvent;
+import org.glassfish.grizzly.IOEventLifeCycleListener;
+import org.glassfish.grizzly.strategies.AbstractIOStrategy;
+
+/**
+ * Grizzly IO Strategy that will handle work using a Mule {@link org.mule.api.context.WorkManager}.
+ */
+public class FlowWorkManagerIOStrategy extends AbstractIOStrategy
```
Comment (andresgregoire) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: Add private constructor

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        return true;
+    }
+
+    @Override
+    public Executor getThreadPoolFor(Connection connection, IOEvent ioEvent)
+    {
+        if (WORKER_THREAD_EVENT_SET.contains(ioEvent))
+        {
+            AsyncHandler handler = ((HttpTransactionContext) connection.getAttributes()
+                    .getAttribute("com.ning.http.client.providers.grizzly.HttpTransactionContext")).getAsyncHandler();
```
Comment (andresgregoire) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: Replace string with HttpTransactionContext.class.getName()

File: modules/http/src/main/java/com/ning/http/client/providers/grizzly/HttpTransactionContext.java (hunk 3 above)
Comment (andresgregoire) [inline modules/http/src/main/java/com/ning/http/client/providers/grizzly/HttpTransactionContext.java]: Sorry, just realized it is a modified version of a class from AHC

File: core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java (hunk 1 above)
Comment (dfeist) [inline core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java]: No, removed.

File: core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java (hunk 2 above)
Comment (dfeist) [inline core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java]: I didn't change format to not add noise to PR as this was just moved.  Can update if required.  You are referring to lack of braces, right?

File: modules/http/src/main/java/org/mule/module/http/internal/request/DefaultHttpRequester.java (hunk 4 above)
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/request/DefaultHttpRequester.java]: Done

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java (hunk 6 above)
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: Done

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java (hunk 5 above)
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: This implementation actually depends on use of grizzly logger.

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java (hunk 7 above)
Comment (dfeist) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/FlowWorkManagerIOStrategy.java]: Opps, can't believe i left this in.  Done.

File: core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+import org.mule.api.processor.ProcessingStrategy;
+import org.mule.api.processor.StageNameSource;
+import org.mule.config.ChainedThreadingProfile;
+import org.mule.processor.AsyncInterceptingMessageProcessor;
+import org.mule.util.concurrent.ThreadNameHelper;
+import org.mule.work.MuleWorkManager;
+
+import java.util.List;
+
+public abstract class AbstractThreadingProfileProcessingStrategy implements ProcessingStrategy
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/processor/strategy/AbstractThreadingProfileProcessingStrategy.java]: can you add some javadoc?

File: core/src/main/java/org/mule/processor/strategy/NonBlockingProcessingStrategy.java [hunk 9]
```diff
@@ -29,8 +21,33 @@
  *
  * @since 3.7
  */
-public class NonBlockingProcessingStrategy extends SynchronousProcessingStrategy
+public class NonBlockingProcessingStrategy extends AbstractThreadingProfileProcessingStrategy
 {
 
+    public static final int DEFAULT_MAX_THREADS = 128;
```
Comment (pablolagreca) [inline core/src/main/java/org/mule/processor/strategy/NonBlockingProcessingStrategy.java]: Can you make this constant private?

File: modules/http/src/main/java/org/mule/module/http/internal/request/HttpClient.java [hunk 10]
```diff
@@ -30,6 +31,6 @@
     /**
      * Sends a HttpRequest without blocking the current thread.  When a response is available or the request times out the provided CompletionHandler will be invoked.
      */
-    public void send(HttpRequest request, int responseTimeout, boolean followRedirects, HttpRequestAuthentication authentication, final CompletionHandler<HttpResponse, Exception> handler);
+    public void send(HttpRequest request, int responseTimeout, boolean followRedirects, HttpRequestAuthentication authentication, final CompletionHandler<HttpResponse, Exception> handler, WorkManager workManager);
```
Comment (pablolagreca) [inline modules/http/src/main/java/org/mule/module/http/internal/request/HttpClient.java]: Can you set a more meaningful name for the workManager?. Something like responseProcessingWorkManager? or Javadoc ?

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/GrizzlyHttpClient.java [hunk 11]
```diff
@@ -251,6 +240,39 @@ public void onThrowable(Throwable t)
         }
     }
 
+    class WorkManagerSourceAsyncCompletionHandler extends AsyncCompletionHandler<Response> implements WorkManagerSource
+    {
+
+        CompletionHandler<HttpResponse, Exception> completionHandler;
```
Comment (pablolagreca) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/GrizzlyHttpClient.java]: Can this variable be private?

File: modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/GrizzlyHttpClient.java [hunk 12]
```diff
@@ -251,6 +240,39 @@ public void onThrowable(Throwable t)
...
+        WorkManager workManager;
```
Comment (pablolagreca) [inline modules/http/src/main/java/org/mule/module/http/internal/request/grizzly/GrizzlyHttpClient.java]: Can this variable be private?

**Pull request closed without merging**